dotnet/src/InternalUtilities/src/Text/StreamJsonParser.cs

```diff
@@ -0,0 +1,222 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.Json.Nodes;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.SemanticKernel.Text;
+
+#pragma warning disable CA1812 // Internal class that is apparently never instantiated
+#pragma warning disable CA1846 // Prefer 'AsSpan' over 'Substring' when span-based overloads are available
+
+/// <summary>
+/// Internal class for parsing a stream of text which contains a series of discrete JSON strings into en enumerable containing each separate JSON string.
+/// </summary>
+/// <remarks>
+/// This class is thread-safe.
+/// </remarks>
+internal sealed class StreamJsonParser
+{
+    /// <summary>
+    /// Parses a Stream containing JSON data and yields the individual JSON objects.
+    /// </summary>
+    /// <param name="stream">The Stream containing the JSON data.</param>
+    /// <param name="validateJson">Set to true to enable checking json chunks are well-formed. Default is false.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>An enumerable collection of string representing the individual JSON objects.</returns>
+    /// <remarks>Stream will be disposed after parsing.</remarks>
+    public async IAsyncEnumerable<string> ParseAsync(
```

Comment:
Maybe simplify the implementation to something more along the lines of this?
https://github.com/dotnet/runtime/issues/33030#issuecomment-1524227075

Reply:
@stephentoub I checked this one and performance is similar, your solution even use less memory, so I eat my words, but it works only with cleaned input, so we can't use this with sse.
https://github.com/Krzysztof318/StreamParserBenchmark

Reply:
Thanks.

> but it works only with cleaned input

Can you elaborate? What can't that approach work with?

Reply:
@stephentoub if input contains sse data like  `event: any-context\ndata: { "person": { "name": "example" } }` or any other not json data like `data: { "person": { "name": "example" } } data done` then `JsonDocument.TryParseValue` will fail